File: tcpHeader.py
import struct
import socket

def getHeader(packet,start):
    tcp = packet[start:start+20] #포트정보를 위한 4byte 잘라냄

    tcp_arr = struct.unpack('!HHLLBBHHH',tcp) #포트번호 파싱
    
    src = tcp_arr[0]
    dst = tcp_arr[1]#포트정보 파싱
        
    tcp_len = (tcp_arr[4] >> 4) * 4 #상위 4bit(길이정보)를 구하고 * 4 하여 최종 길이 산출 

    return src,dst,start + tcp_len #포트정보 반환

File: tcpHeader.c
// tcpHeader.c
#include <string.h>
#include <arpa/inet.h>
#include "tcpHeader.h"

int getTcpHeader(const u_char *packet, int start, int *src_port, int *dst_port) {
    const u_char *tcp = packet + start;

    uint16_t src, dst;
    uint8_t offset;

    memcpy(&src, tcp, 2);
    memcpy(&dst, tcp + 2, 2);
    memcpy(&offset, tcp + 12, 1);

    *src_port = ntohs(src);
    *dst_port = ntohs(dst);

    int tcp_len = (offset >> 4) * 4;

    return start + tcp_len;
}
